src/coreclr/jit/inductionvariableopts.cpp

```diff
@@ -0,0 +1,1533 @@
+            break;
+        }
+        case ScevOper::Local:
+        {
+            ScevLocal* invariantLocal = (ScevLocal*)scev;
+            printf("V%02u.%u", invariantLocal->LclNum, invariantLocal->SsaNum);
+
+            int64_t cns;
+            if (invariantLocal->GetConstantValue(m_comp, &cns))
+            {
+                printf(" (%lld)", (long long)cns);
+            }
+            break;
+        }
+        case ScevOper::ZeroExtend:
+        case ScevOper::SignExtend:
+        {
+            ScevUnop* unop = (ScevUnop*)scev;
+            printf("%cext<%d>(", unop->Oper == ScevOper::ZeroExtend ? 'z' : 's', genTypeSize(unop->Type) * 8);
+            DumpScev(unop->Op1);
+            printf(")");
+            break;
+        }
+        case ScevOper::Add:
+        case ScevOper::Mul:
+        case ScevOper::Lsh:
+        {
+            ScevBinop* binop = (ScevBinop*)scev;
+            printf("(");
+            DumpScev(binop->Op1);
+            const char* op;
+            switch (binop->Oper)
+            {
+                case ScevOper::Add:
+                    op = "+";
+                    break;
+                case ScevOper::Mul:
+                    op = "*";
+                    break;
+                case ScevOper::Lsh:
+                    op = "<<";
+                    break;
+                default:
+                    unreached();
+            }
+            printf(" %s ", op);
+            DumpScev(binop->Op2);
+            printf(")");
+            break;
+        }
+        case ScevOper::AddRec:
+        {
+            ScevAddRec* addRec = (ScevAddRec*)scev;
+            printf("<" FMT_LP, m_loop->GetIndex());
+            printf(", ");
+            DumpScev(addRec->Start);
+            printf(", ");
+            DumpScev(addRec->Step);
+            printf(">");
+            break;
+        }
+        default:
+            unreached();
+    }
+}
+#endif
+
+//------------------------------------------------------------------------
+// CreateSimpleInvariantScev: Create a "simple invariant" SCEV node for a tree:
+// either an invariant local use or a constant.
+//
+// Parameters:
+//   tree - The tree
+//
+// Returns:
+//   SCEV node or nullptr if the tree is not a simple invariant.
+//
+Scev* ScalarEvolutionContext::CreateSimpleInvariantScev(GenTree* tree)
+{
+    if (tree->OperIs(GT_CNS_INT, GT_CNS_LNG))
+    {
+        return CreateScevForConstant(tree->AsIntConCommon());
+    }
+
+    if (tree->OperIs(GT_LCL_VAR) && tree->AsLclVarCommon()->HasSsaName())
+    {
+        LclVarDsc*    dsc    = m_comp->lvaGetDesc(tree->AsLclVarCommon());
+        LclSsaVarDsc* ssaDsc = dsc->GetPerSsaData(tree->AsLclVarCommon()->GetSsaNum());
+
+        if ((ssaDsc->GetBlock() == nullptr) || !m_loop->ContainsBlock(ssaDsc->GetBlock()))
+        {
+            return NewLocal(tree->AsLclVarCommon()->GetLclNum(), tree->AsLclVarCommon()->GetSsaNum());
+        }
+    }
+
+    return nullptr;
+}
```

Comment:
Hoisting has a more refined notion of loop invariance that can handle more tree shapes; any thought of using something like that here?

Reply:
Do you mean by switching to use VN? I suppose that would be possible (although somewhat annoying since the VN wouldn't give me things like "reaching def from outside the loop" that I'm making use of here). I think it would mainly get us the ability to reason about loop invariant loads (so you could have e.g. `<L00, 0, IND(x)>`, provided we also started representing loads in the SCEV IR).